Allow device cluster entities overwrite from v2 quirks

mrrstux opened the pull request:
Entities created by a v2 quirk should be allowed to overwrite any default generated ones.

Current patch harmonizes entities unique ids format across the component. It requires HA entities unique id migration code.

Fixes https://github.com/home-assistant/core/issues/128253.

mrrstux committed:
Allow device cluster entities overwrite from v2 quirks.

dmulcahey commented on zha/application/discovery.py:
```diff
@@ -314,7 +314,7 @@ def discover_quirks_v2_entities(self, device: Device) -> None:
                     platform=platform,
                     entity_class=entity_class,
-                    unique_id=endpoint.unique_id,
+                    unique_id=f"{endpoint.unique_id}-{cluster.cluster_id}",
```
I think this will break all existing v2 entities and it will require a migration of the HA entity registry.

mrrstux replied:
Is that migration a post-install operation?

mrrstux replied:
In ZHA `async_setup_entry` I can get access to all HA ZHA entities. But it doesn't look like the HA associated entity data contains the info I need to migrate the ids - I would not know the cluster id of each entity in order to migrate to `<unique id>-<cluster id>`.

Here is how the registry entry looks like for a local temperature calibration entity:

```
'RegistryEntry(entity_id='number.thermostat_trv_local_temperature_offset', unique_id='28:fa:26:01:02:0b:67:2a-1-513-local_temperature_calibration', platform='zha', previous_unique_id=None, aliases=set(), area_id=None, categories={}, capabilities={'min': -5.0, 'max': 5.0, 'step': 0.1, 'mode': 'auto'}, config_entry_id='75fe69aebaacaaed3a012a97167d2d10', created_at=datetime.datetime(1970, 1, 1, 0, 0, tzinfo=datetime.timezone.utc), device_class=None, device_id='9e9268b18b27e97adb5d7543c1df8f62', disabled_by=None, entity_category=<EntityCategory.CONFIG: 'config'>, hidden_by=None, icon=None, id='617948e1ccca4e53075e64956998a17e', has_entity_name=True, labels=set(), modified_at=datetime.datetime(2024, 12, 16, 19, 51, 43, 399371, tzinfo=datetime.timezone.utc), name=None, options={'conversation': {'should_expose': False}}, original_device_class=None, original_icon=None, original_name='Local temperature offset', supported_features=0, translation_key='local_temperature_calibration', unit_of_measurement='°C', _cache={})'.
```

Is there any other place where the migration could be done?

mrrstux replied:
Found a way: inside BaseEntityInfo, the previous unique id can be stored and used by ZHA to migrate existing HA entities.

mrrstux replied:
Here is the migration code: https://github.com/home-assistant/core/pull/136788.

dmulcahey commented on zha/application/platforms/__init__.py:
```diff
@@ -310,15 +310,14 @@ def __init__(
+                "Duplicate entity detected - unique id %r already exists: %r. Replacing with %r",
                 (self.PLATFORM, self.unique_id),
                 self._device.platform_entities[(self.PLATFORM, self.unique_id)],
+                self,
```
What is this doing?

mrrstux replied:
Intention was to offer a debugging message for those cases where it is not intended or expected to encounter duplicated entities. Can be removed or demoted to debug.

dmulcahey commented on zha/application/platforms/__init__.py:
```diff
@@ -310,15 +310,14 @@ def __init__(
                 self._device.platform_entities[(self.PLATFORM, self.unique_id)],
+                self,
             )
+        self._device.platform_entities[(self.PLATFORM, self.unique_id)] = self
```
Same question as above

mrrstux replied:
It is allowing a newer entity instance to overwrite an existing one. In my specific case, an automatic zha created local temp calibration entity would be replaced by one provided by a quirk v2. At least, in my setup, the quirk v2 entity gets created last.

mrrstux committed:
Fix tests.

mrrstux committed:
Transition all entities to new unique id scheme

<endpoint unique id>-<cluster id>-<suffix>

Keep track of previous unique id inside BaseEntityInfo to allow
home assistant zha component to migrate existing entities.

puddly committed:
Merge branch 'dev' into puddly/mrrstux/quirk2-allow-entity-override

puddly committed:
Reorganize entity creation to rely less on shared state

puddly committed:
Migrate everything to use unique ID suffixes

puddly commented:
@mrrstux I'm going to pick this PR up to rebase it on top of `dev` and then deal with a bunch of corner cases for migration. Do you mind if I push commits straight to your branch? Or would you prefer I make a separate PR to supersede this one?

puddly committed:
Begin fixing unit tests

puddly committed:
Generalize unique ID roots to be more explicit

mrrstux commented:
@puddly Yes, by all means, please take it over!

puddly committed:
Allow entity ID migration to be more generic

puddly committed:
WIP: Cleanup, preparation for unused migration feature removal

puddly committed:
Drop `LEGACY_DISCOVERY` in favor of discovery-less migration

puddly committed:
Hard-code the current unique ID suffix

puddly committed:
Migrate automatically by default

puddly committed:
Fix existing platforms

puddly committed:
Drop `legacy_discovery_unique_id`

puddly committed:
Delete an invalid entity from a failing device

puddly committed:
Fix most existing unit tests

puddly committed:
Do not replace current entities when re-initializing

puddly committed:
Migrate quirks v2 entities

puddly committed:
Migrate endpoint-level climate entities as well

puddly committed:
Clean up discovery code slightly

puddly committed:
Fix unit test referencing v2 quirk

puddly committed:
Merge branch 'dev' into quirk2-allow-entity-override

puddly committed:
Clean up a few more dynamically-computed `_unique_id_suffix` attrs

puddly committed:
Merge branch 'dev' into quirk2-allow-entity-override

puddly committed:
Fix unit tests after merge

puddly committed:
Exclude internal helper function error cases from coverage

puddly commented:
I think this PR is ready to go for the upcoming beta.

This will affect every ZHA entity in Home Assistant so I think the Home Assistant side of this change needs a bit of a rework and to pre-compute the migration in order to perform it pseudo-atomically.

puddly committed:
Revert all migrations, hand off to a second PR

puddly committed:
Merge branch 'dev' into quirk2-allow-entity-override

puddly commented:
I've scaled back the changes in this PR to *not* perform any migrations: it only affects the way entities are created and adds a new attribute to entity info to hold migrations. The format for it is provisional and likely will change (likely to include a version prefix for performance reasons?).
